File: microservice-architecture/rate-limiting/go/src/limiter.go
// Package src 实现了限流模式（Rate Limiting）的核心逻辑。
//
// 【设计模式】策略模式：固定窗口是一种限流策略。
//
// 【架构思想】限流保护系统不被过载流量拖垮。
//
// 【开源对比】
//   - uber-go/ratelimit：令牌桶限流器
//   - golang.org/x/time/rate：Go 标准扩展库的令牌桶
//   本示例实现固定窗口，省略了时间和并发安全。
package src

// FixedWindowLimiter 固定窗口限流器。
// 【设计模式】策略模式：固定窗口是一种限流策略。
type FixedWindowLimiter struct {
	limit int // 窗口内最大允许请求数
	count int // 当前窗口已通过请求数
}

// NewFixedWindowLimiter 创建固定窗口限流器。
func NewFixedWindowLimiter(limit int) *FixedWindowLimiter {
	return &FixedWindowLimiter{limit: limit}
}

// Allow 判断是否允许通过。count < limit 时放行，否则拒绝。
func (l *FixedWindowLimiter) Allow() bool {
	// 达到上限，拒绝
	if l.count >= l.limit {
		return false
	}
	// 放行并递增
	l.count++
	return true
}

// AdvanceWindow 推进窗口，重置计数。
func (l *FixedWindowLimiter) AdvanceWindow() {
	l.count = 0
}

File: microservice-architecture/rate-limiting/c/src/limiter.c
/**
 * @file limiter.c - 限流模式（Rate Limiting）的 C 语言实现
 *
 * 【设计模式】策略模式：固定窗口是一种限流策略。
 *
 * 【架构思想】
 *   C 语言中限流通常嵌入在网络框架中。Nginx 的 limit_req 模块
 *   使用漏桶算法实现请求限流。
 *
 * 【开源对比】
 *   - Nginx limit_req：漏桶算法 + burst 突发允许
 *   - Envoy：支持本地限流和全局限流（ratelimit service）
 *   本示例实现最简单的固定窗口。
 */

#include "func.h"

/** 初始化限流器 */
void limiter_init(FixedWindowLimiter *limiter, int limit)
{
    limiter->limit = limit;
    limiter->count = 0;
}

/**
 * 判断是否允许通过。
 * count < limit 时放行并递增，否则拒绝。
 * @return 1=放行，0=拒绝
 */
int limiter_allow(FixedWindowLimiter *limiter)
{
    /* 达到上限，拒绝 */
    if (limiter->count >= limiter->limit) {
        return 0;
    }
    /* 放行并递增 */
    limiter->count++;
    return 1;
}

/** 推进窗口，重置计数 */
void limiter_advance_window(FixedWindowLimiter *limiter)
{
    limiter->count = 0;
}
